### DISCUSSION TITLE
Can I put the props of an island in a separate js file instead of the HTML file?

### DISCUSSION THREAD

**Discussion by ouuan** [ElMassimo/iles#202] [Q&A] [answered]
I have a [search page](https://github.com/ouuan/iles-blog/blob/1d1c4eef52ca4765d00bbfd033e31f8b3d367d6f/src/pages/search.vue) in my blog. Now I calculate the props for the search component outside of the island so that it runs on the server side. But the props are placed in the HTML file, so it's getting bigger and bigger. I want to put the props in a separate `.js` file. How can I do that?

Alternatively, is it possible to get the data of `useDocuments` outside of iles, for example, in a Vite plugin, so that I can generate the search data elsewhere?

**Comment by ElMassimo** [comment, marked as answer]
Hi Yufan!

This is an interesting use case. In îles props are inlined in HTML, there's no built-in way to store them as JS instead.

You can get the data of `useDocuments` using an internal API, by adding an import like the following inside a script that you import from the island:

```ts
import documents from '/@islands/documents?pattern=~/pages/{post/**/*,about}.{md,mdx}'
```

That way, that data would be stored in JS, and you could even use a dynamic import to load it asynchronously.

---

If you want to get more control by creating a Vite plugin, you can [use the `pages` API](https://github.com/ElMassimo/iles/blob/main/packages/iles/src/node/plugin/documents.ts#LL59) to get the `frontmatter`, `meta`, and `excerpt`, and [read the raw file](https://github.com/ElMassimo/iles/blob/main/packages/iles/src/node/plugin/documents.ts#L60) as you are [doing now](https://github.com/ouuan/iles-blog/blob/master/src/pages/search.vue#L22).

The downside of this search technique, is that the browser needs to download the entire content of site.

---

[In my blog](https://maximomussini.com/) and in the îles docs I'm [using DocSearch to power search](https://docsearch.algolia.com/), and have a [CI task](https://github.com/ElMassimo/iles/blob/main/.github/workflows/build.yml#L117-L135) that will crawl the site every time it's deployed to make sure the indexes are up-to-date.

There are other alternatives like this, which are very efficient and work nicely even in mobile devices and slow network conditions.
